### PR TITLE
Extend client area to decorations API

### PR DESCRIPTION
**Pull request by danwalmsley**
## What does the pull request do?

Depends on #3962 

dwm-api

This adds an API to allow drawing inside the titlebar, and having custom window decorations while maintaining native window frame (on windows)

It adds the following Hint properties:

`ExtendClientAreaToDecorationsHint`, `ExtendClientAreaChromeHints` and `ExtendClientAreaTitleBarHeightHint`

Once enabled the following property tells the user the platform supported extending:
`IsExtendedIntoWindowDecorations`

If enabled then the following properties can be used to control the layout.
`WindowDecorationMargins` this tells you the margin you need around your content area so that it wont overlap the window border or titlebar area.

`OffScreenMargin` this tells you the amount the window is offscreen (as is the case on windows then maximized. It is not normally needed by end user, but it is used in the Window template.

## What is the current behavior?

Currently its:
1) not possible to put content in titlebar.
2) quite difficult to customize the Window and have it behave correctly on various platforms.

## What is the updated/expected behavior with this PR?

This will allow users to put content in the titlebar as required, and fully customize the window correctly.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

**Review by FrankenApps — APPROVED** [review]

**Comment by FrankenApps** [comment]
First of all great work! 
I have only quickly checked the Control Gallery on Windows and have noticed some minor issues.

* The **Titlebar** checkbox should probably be disabled, if **Prefer System Chrome** is enabled, because these options influence each other. For example if **Titlebar** is checked, the **Prefer System Chrome** option is inversed. (See picture below: I think the default buttons should be visible there, right?)
* Is it possible to change the (currently black) color of the overlay in the title bar, which hides the default System Chrome Buttons? I tried to find an option, but did not succeed. This would be important, if someone wanted to implement their own title bar with a custom (not system defined) color.
not_transparent

* If the title bar is hidden and the window is blurred or transparent one can still see the highlight color, when hovering the mouse over the default system chrome buttons. E.g. There seems to be a tiny gap, between the overlay and the top border of the window.
blur

**Comment by danwalmsley** [comment]
> First of all great work!
> I have only quickly checked the Control Gallery on Windows and have noticed some minor issues.
> 
> * The **Titlebar** checkbox should probably be disabled, if **Prefer System Chrome** is enabled, because these options influence each other. For example if **Titlebar** is checked, the **Prefer System Chrome** option is inversed. (See picture below: I think the default buttons should be visible there, right?)

that is only the case on windows, on osx you will have a different behavior.

File: src/Avalonia.Themes.Fluent/CaptionButtons.xaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+<Styles xmlns="https://github.com/avaloniaui" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
+  <Style Selector="CaptionButtons">
+    <Setter Property="Foreground" Value="{DynamicResource ThemeForegroundBrush}"/>
+    <Setter Property="Canvas.Right" Value="0" />
```
Comment (grokys) [inline src/Avalonia.Themes.Fluent/CaptionButtons.xaml:4]: Is this necessary? Where does CaptionButtons appear in a canvas?

File: src/Avalonia.Themes.Fluent/CaptionButtons.xaml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          <Panel x:Name="PART_MinimiseButton">
+            <Viewbox>
+              <Path Stretch="UniformToFill" Fill="{TemplateBinding Foreground}" Data="M2048 1229v-205h-2048v205h2048z" />
+            </Viewbox>
+          </Panel>
+
+          <Panel x:Name="PART_RestoreButton">
+            <Viewbox>
+              <Viewbox.RenderTransform>
+                <RotateTransform Angle="-90" />
```
Comment (grokys) [inline src/Avalonia.Themes.Fluent/CaptionButtons.xaml:40]: Why is this needed? Restore button path seems to be a square.

File: src/Avalonia.Themes.Default/TitleBar.xaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  </Design.PreviewWith>
+  <Style Selector="TitleBar">
+    <Setter Property="Foreground" Value="{DynamicResource SystemControlForegroundBaseHighBrush}"/>    
+    <Setter Property="VerticalAlignment" Value="Top" />
+    <Setter Property="HorizontalAlignment" Value="Stretch" />
+    <Setter Property="Background" Value="Transparent" />
+    <Setter Property="Template">
+      <ControlTemplate>
+        <Panel HorizontalAlignment="{TemplateBinding HorizontalAlignment}" VerticalAlignment="Stretch">
+          <Panel x:Name="PART_MouseTracker" Height="1" VerticalAlignment="Top" />
```
Comment (grokys) [inline src/Avalonia.Themes.Default/TitleBar.xaml:15]: I can't see `PART_MouseTracker` referred to in code, is this still needed? Or `PART_Container` or `PART_Background` for that matter.

**Review by grokys — CHANGES_REQUESTED** [review]
A few things:

- It seems strange to me that the Titlebar property only has effect when the window is transparent, those two things seem to be unrelated to me?
- For the first release are we going to be showing Windows chrome on all platforms? (Fine if so, just wanted to check)
- There are a few potential problems with null references. Can we add `#nullable enable` to the new files at least?
- I mention it in a few places, but XML docs for the new APIs would be useful
- Fluent dark theme seems to be broken on this branch?

File: src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+                            Resized(clientSize / Scaling);
+                        }
+
+                        var windowState = size == SizeCommand.Maximized ?
+                            WindowState.Maximized :
+                            (size == SizeCommand.Minimized ? WindowState.Minimized : WindowState.Normal);
+
+                        if (windowState != _lastWindowState)
+                        {
+                            _lastWindowState = windowState;
```
Comment (yahiheb) [inline src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs:380]:
```suggestion
                            _lastWindowState = windowState;
```

**Review by yahiheb — CHANGES_REQUESTED** [review]

File: src/Avalonia.Controls/Window.cs [hunk 5]
```diff
@@ -435,6 +534,29 @@ protected virtual void HandleWindowStateChanged(WindowState state)
...
 
+        protected virtual void ExtendClientAreaToDecorationsChanged(bool isExtended)
+        {
+            IsExtendedIntoWindowDecorations = isExtended;
+
+            WindowDecorationMargins = PlatformImpl.ExtendedMargins;
+
+            OffScreenMargin = PlatformImpl.OffScreenMargin;
+
+            if(PlatformImpl.NeedsManagedDecorations)
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Window.cs:545]: fixed

File: src/Avalonia.Controls/Window.cs [hunk 6]
```diff
@@ -237,6 +290,52 @@ public bool HasSystemDecorations
...
+        /// <summary>
+        /// Gets if the ClientArea is Extended into the Window Decorations.
+        /// </summary>
+        public bool IsExtendedIntoWindowDecorations
+        {
+            get => _isExtendedIntoWindowDecorations;
+            private set => SetAndRaise(IsExtendedIntoWindowDecorationsProperty, ref _isExtendedIntoWindowDecorations, value);
+        }
+
+        private Thickness _windowDecorationMargins;
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Window.cs:323]: fixed.

File: src/Avalonia.Controls/Chrome/CaptionButtons.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+        {
+            base.OnApplyTemplate(e);
+
+            var closeButton = e.NameScope.Find<Panel>("PART_CloseButton");
+            var restoreButton = e.NameScope.Find<Panel>("PART_RestoreButton");
+            var minimiseButton = e.NameScope.Find<Panel>("PART_MinimiseButton");
+            var fullScreenButton = e.NameScope.Find<Panel>("PART_FullScreenButton");
+
+            closeButton.PointerReleased += (sender, e) => _hostWindow.Close();
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Chrome/CaptionButtons.cs:57]: done, used Get.

File: src/Avalonia.Controls/Chrome/CaptionButtons.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                    })
+                };
+            }
+        }
+
+        public void Detach()
+        {
+            if (_disposables != null)
+            {
+                var layer = ChromeOverlayLayer.GetOverlayLayer(_hostWindow);
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Chrome/CaptionButtons.cs:39]: done

File: src/Avalonia.Controls/Chrome/TitleBar.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+                _captionButtons?.Detach();
+            }
+        }
+
+        protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+        {
+            base.OnApplyTemplate(e);
+
+            _captionButtons = e.NameScope.Find<CaptionButtons>("PART_CaptionButtons");
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Chrome/TitleBar.cs:96]: done

File: src/Avalonia.Controls/Chrome/TitleBar.cs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+                    })
+                };
+
+                _captionButtons?.Attach(_hostWindow);
+
+                InvalidateSize();
+            }
+        }
+
+        void InvalidateSize()
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Chrome/TitleBar.cs:58]: done.

File: src/Avalonia.Controls/Chrome/TitleBar.cs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            if (_disposables == null)
+            {
+                var layer = ChromeOverlayLayer.GetOverlayLayer(_hostWindow);
+
+                layer.Children.Add(this);
+
+                _disposables = new CompositeDisposable
+                {
+                    _hostWindow.GetObservable(Window.WindowDecorationMarginsProperty)
+                    .Subscribe(x => InvalidateSize()),
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Chrome/TitleBar.cs:34]: done.

File: src/Avalonia.Controls/Platform/IWindowImpl.cs [hunk 12]
```diff
@@ -94,5 +94,51 @@ public interface IWindowImpl : IWindowBaseImpl
...
+        /// <summary>
+        /// Sets if the ClientArea is extended into the non-client area.
+        /// </summary>
+        /// <param name="extendIntoClientAreaHint">true to enable, false to disable</param>
+        void SetExtendClientAreaToDecorationsHint(bool extendIntoClientAreaHint);
+
+        /// <summary>
+        /// Gets a value to indicate if the platform was able to extend client area to non-client area.
+        /// </summary>
+        bool IsClientAreaExtendedToDecorations { get; }
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Platform/IWindowImpl.cs:107]: done.

File: src/Avalonia.Controls/Chrome/CaptionButtons.cs [hunk 13]
```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reactive.Disposables;
+using System.Text;
+using Avalonia.Controls.Primitives;
+using Avalonia.VisualTree;
+
+namespace Avalonia.Controls.Chrome
+{
+    public class CaptionButtons : TemplatedControl
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Chrome/CaptionButtons.cs:12]: done

File: src/Avalonia.Controls/Chrome/TitleBar.cs [hunk 14]
```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Reactive.Disposables;
+using Avalonia.Controls.Primitives;
+
+namespace Avalonia.Controls.Chrome
+{
+    public class TitleBar : TemplatedControl
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Chrome/TitleBar.cs:12]: done

File: src/Avalonia.Controls/Chrome/TitleBar.cs [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        private Window _hostWindow;
+        private CaptionButtons _captionButtons;
+
+        public TitleBar(Window hostWindow)
+        {
+            _hostWindow = hostWindow;
+        }
+
+        public TitleBar()
+        {
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Chrome/TitleBar.cs:24]: added nullable and null checking.

File: src/Avalonia.Controls/Window.cs [hunk 16]
```diff
@@ -435,6 +534,29 @@ protected virtual void HandleWindowStateChanged(WindowState state)
             }
         }
 
+        protected virtual void ExtendClientAreaToDecorationsChanged(bool isExtended)
+        {
+            IsExtendedIntoWindowDecorations = isExtended;
+
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Window.cs:537]: done

File: src/Avalonia.Themes.Default/CaptionButtons.xaml [hunk 17]
```diff
@@ -0,0 +1,70 @@
+<Styles xmlns="https://github.com/avaloniaui" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
+  <Style Selector="CaptionButtons">
+    <Setter Property="Foreground" Value="{DynamicResource ThemeForegroundBrush}"/>
+    <Setter Property="Canvas.Right" Value="0" />
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Themes.Default/CaptionButtons.xaml:4]: done.

File: src/Avalonia.Themes.Fluent/CaptionButtons.xaml (hunk 1 above)
Comment (danwalmsley) [reply src/Avalonia.Themes.Fluent/CaptionButtons.xaml:4]: done.

File: src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs (hunk 4 above)
Comment (danwalmsley) [reply src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs:380]: done.

File: src/Avalonia.Native/WindowImplBase.cs [hunk 18]
```diff
@@ -266,6 +266,11 @@ public bool RawKeyEvent(AvnRawKeyEventType type, uint timeStamp, AvnInputModifie
             return args.Handled;
         }
 
+        protected virtual bool ChromeHitTest (RawPointerEventArgs e)
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Native/WindowImplBase.cs:269]: done;

File: src/Avalonia.Native/WindowImpl.cs [hunk 19]
```diff
@@ -96,7 +102,85 @@ public WindowState WindowState
...
+                    }
+                }
+            }
+
+            return false;
+        }
+        
+        private void InvalidateExtendedMargins ()
+        {
+            if(WindowState ==  WindowState.FullScreen)
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Native/WindowImpl.cs:143]: done.

File: src/Avalonia.Native/WindowImpl.cs [hunk 20]
```diff
@@ -96,7 +102,85 @@ public WindowState WindowState
...
+                    {
+                        _native.BeginMoveDrag();
+                    }
+                }
+            }
+
+            return false;
+        }
+        
+        private void InvalidateExtendedMargins ()
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Native/WindowImpl.cs:141]: done.

File: src/Avalonia.Controls/Window.cs [hunk 21]
```diff
@@ -87,6 +91,39 @@ public class Window : WindowBase, IStyleable, IFocusScope, ILayoutRoot
...
+        /// </summary>
+        public static readonly DirectProperty<Window, Thickness> WindowDecorationMarginsProperty =
+            AvaloniaProperty.RegisterDirect<Window, Thickness>(nameof(WindowDecorationMargins),
+                o => o.WindowDecorationMargins);
+
+        public static readonly DirectProperty<Window, Thickness> OffScreenMarginProperty =
+            AvaloniaProperty.RegisterDirect<Window, Thickness>(nameof(OffScreenMargin),
+                o => o.OffScreenMargin);
+
+
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Window.cs:127]: done.

File: src/Avalonia.Controls/Window.cs [hunk 22]
```diff
@@ -87,6 +91,39 @@ public class Window : WindowBase, IStyleable, IFocusScope, ILayoutRoot
...
+        /// </summary>
+        public static readonly StyledProperty<bool> ExtendClientAreaToDecorationsHintProperty =
+            AvaloniaProperty.Register<Window, bool>(nameof(ExtendClientAreaToDecorationsHint), false);
+
+        public static readonly StyledProperty<ExtendClientAreaChromeHints> ExtendClientAreaChromeHintsProperty =
+            AvaloniaProperty.Register<Window, ExtendClientAreaChromeHints>(nameof(ExtendClientAreaChromeHints), ExtendClientAreaChromeHints.Default);
+
+        public static readonly StyledProperty<double> ExtendClientAreaTitleBarHeightHintProperty =
+            AvaloniaProperty.Register<Window, double>(nameof(ExtendClientAreaTitleBarHeightHint), -1);
+
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Window.cs:106]: done.

File: samples/ControlCatalog/Pages/WindowCustomizationsPage.xaml [hunk 23]
```diff
@@ -0,0 +1,19 @@
+﻿<UserControl xmlns="https://github.com/avaloniaui"
+             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
+             x:Class="ControlCatalog.Pages.WindowCustomizationsPage">
+  <StackPanel Spacing="10"  Margin="25">
+    <CheckBox Content="Extend Client Area to Decorations" IsChecked="{Binding ExtendClientAreaEnabled}" />
+    <CheckBox Content="Titlebar" IsChecked="{Binding SystemTitleBarEnabled}" />
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply samples/ControlCatalog/Pages/WindowCustomizationsPage.xaml:9]: done.

File: samples/ControlCatalog/MainWindow.xaml [hunk 24]
```diff
@@ -56,20 +61,30 @@
...
+          </MenuItem>
+        </Menu>
+        <local:MainView />
+      </DockPanel>
+    </Panel>
+    <Border IsVisible="{Binding ExtendClientAreaEnabled}" BorderThickness="1 1 1 0" CornerRadius="4 4 0 0" BorderBrush="#55000000" Height="22" VerticalAlignment="Top" HorizontalAlignment="Left" Margin="250 8 0 0">
+      <Border.Background>
+        <SolidColorBrush Color="White" Opacity="0.7" />
+      </Border.Background>
+      <TextBlock Margin="5 5 5 0" Text="Content In TitleBar" />
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply samples/ControlCatalog/MainWindow.xaml:87]: done.

File: src/Avalonia.Controls/Primitives/ChromeOverlayLayer.cs [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+            return null;
+        }
+
+        public bool HitTest(Point point) => Children.HitTestCustom(point);
+
+        protected override Size ArrangeOverride(Size finalSize)
+        {
+            // We are saving it here since child controls might need to know the entire size of the overlay
+            // and Bounds won't be updated in time
+            AvailableSize = finalSize;
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Primitives/ChromeOverlayLayer.cs:33]: removed.

File: src/Avalonia.Controls/Platform/ExtendClientAreaChromeHints.cs [hunk 26]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+
+namespace Avalonia.Platform
+{
+    [Flags]
+    public enum ExtendClientAreaChromeHints
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Controls/Platform/ExtendClientAreaChromeHints.cs:6]: done.

File: src/Avalonia.Themes.Fluent/CaptionButtons.xaml (hunk 2 above)
Comment (danwalmsley) [reply src/Avalonia.Themes.Fluent/CaptionButtons.xaml:40]: image
not when maximised.

File: src/Avalonia.Themes.Fluent/CaptionButtons.xaml [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+              </Viewbox.RenderTransform>
+              <Path Stretch="UniformToFill" Fill="{TemplateBinding Foreground}"/>
+            </Viewbox>
+          </Panel>
+
+          <Panel x:Name="PART_CloseButton">
+            <Viewbox>
+              <Path Stretch="UniformToFill" Fill="{TemplateBinding Foreground}" Data="M1169 1024l879 -879l-145 -145l-879 879l-879 -879l-145 145l879 879l-879 879l145 145l879 -879l879 879l145 -145z" />
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Themes.Fluent/CaptionButtons.xaml:48]: Close button never changes, the others do depending on state.

File: src/Avalonia.Themes.Default/TitleBar.xaml (hunk 3 above)
Comment (danwalmsley) [reply src/Avalonia.Themes.Default/TitleBar.xaml:15]:
```
<Style Selector="TitleBar:fullscreen /template/ Panel#PART_MouseTracker">
    <Setter Property="Background" Value="Transparent" />
  </Style>
```

it works by becoming transparent in fullscreen mode... so it means the titlebar gets a pointerover applied...

that then triggers the animation to reveal the titlebar.

Comment (danwalmsley) [reply src/Avalonia.Themes.Default/TitleBar.xaml:15]: Styles are referencing the other PARTs also.

**Comment by danwalmsley** [comment]
> * It seems strange to me that the Titlebar property only has effect when the window is transparent, those two things seem to be unrelated to me?
Titlebar is there, but the user need to not render over that area to see it on windows.. there is a margin they can bind to to do this.. or they need to render transparent content there... 

image
Edge here its tabs are actually transparent so the titlebar can be seen.

I will improve the control catalog code so that it demonstrates this better.

> * For the first release are we going to be showing Windows chrome on all platforms? (Fine if so, just wanted to check)
Yes, OSX we will never use managed chrome as OSX thankfully draws the chrome buttons on top of the user content. Apple 1, Msft 0.

Linux, there is no standard chrome, however I propose to follow this up with a PR that allows you to template managed chrome.
including resize borders.

**Comment by danwalmsley** [comment]
@FrankenApps 
If the title bar is hidden and the window is blurred or transparent one can still see the highlight color, when hovering the mouse over the default system chrome buttons. E.g. There seems to be a tiny gap, between the overlay and the top border of the window.

fixed.

**Comment by danwalmsley** [comment]
@grokys iv fixed some of the issues, like the dark / light mode works again.

The native windows titlebar being affected by transparency... isnt really an issue, if the user wants to draw over the titlebar and have the titlebar show through, they need to fill the back buffer with transparent pixels... 

Iv made it so they can now do the following: 
Set `TransparencyBackgroundFallback="Transparent"`
Use `Window.WindowDecorationMargins` property to offset their content to not cover the titlebar.

File: src/Avalonia.Themes.Default/TitleBar.xaml (hunk 3 above)
Comment (grokys) [reply src/Avalonia.Themes.Default/TitleBar.xaml:15]: The `PART_` convention is usually used to mean that a control will be used in the code for the control. I can't actually find an explicit mention of this in the WPF docs, but I can find a few mentions on other sites such as http://bursjootech.blogspot.com/2008/11/naming-and-conventions-in-xaml.html

> "In WPF there is a convention for TemplateElements in controls that distinguishes controls that must exist in a template from other controls. This means that if I edit a control template I know that all elements named PART_* must be in the template and all other parts I can remove or do whatever I like to.

Because these parts aren't part of the contract between the code of the control and the template, they probably shouldn't have the `PART_` prefix. It's not a big deal I guess, but when I saw the `PART_` prefix and no code to get these controls I assumed it was dead code.

**Comment by danwalmsley** [comment]
for the record @kekekeks  and @grokys agreed to do a YOLO on this. since its passed a few review passes now.
